FW3: Revise make system for Unit Tests

StevenClinardSpanIO opened the pull request:
General overhaul of the make system to normalize usage.

Use include.cmake at each primary level and move global include, compile flags, etc. into appropriate include.cmake

Introduce CMakeLists.txt into forked 3rd party repositories to allow a more one-to-one mapping of cmake stuff and content

Move stuff from the "combo" rtos+processor into corresponding one-to-one include.cmake and CMakeLists.txt.
  This reduces the content of the "combo" cmake files (e.g. at FW3-STM32/RTOS/FreeRTOS) to a bare minimum that
    truly depends on the combo.

Make proper targets for the Mock CMSIS, RTOS, CPU

FW3-Components/include.cmake will detect and override the various targets to the Mock targets for unit test

Introduced fw3-add-tests.cmake which defines a single function to create a test binary and add all unit tests for the module
  via gtest_discovery_tests function
  This greatly simplifies the cmake work to integrate tests and will automatically find and add tests.

cmake-format all this stuff

StevenClinardSpanIO committed:
FW3: Revise make system for Unit Tests

General overhaul of the make system to normalize usage.

Use include.cmake at each primary level and move global include, compile flags, etc. into appropriate include.cmake

Introduce CMakeLists.txt into forked 3rd party repositories to allow a more one-to-one mapping of cmake stuff and content

Move stuff from the "combo" rtos+processor into corresponding one-to-one include.cmake and CMakeLists.txt.
  This reduces the content of the "combo" cmake files (e.g. at FW3-STM32/RTOS/FreeRTOS) to a bare minimum that
    truly depends on the combo.

Make proper targets for the Mock CMSIS, RTOS, CPU

FW3-Components/include.cmake will detect and override the various targets to the Mock targets for unit test

Introduced fw3-add-tests.cmake which defines a single function to create a test binary and add all unit tests for the module
  via gtest_discovery_tests function
  This greatly simplifies the cmake work to integrate tests and will automatically find and add tests.

cmake-format all this stuff

zhixuenlai commented on Source/include.cmake:
```diff
@@ -0,0 +1,3 @@
+set(FREERTOS_CONFIG_FILE_DIRECTORY ${FW3_APPLICATION_DIR}/Include)
+
+set(FREERTOS_HEAP ${FW3_COMPONENTS_DIR}/Source/no_heap.c)
```
When I try to build the project locally, it fails because these variables are not set. It turns out that we have to define them in the following ways:
```cmake
set(FREERTOS_CONFIG_FILE_DIRECTORY ${FW3_APPLICATION_DIR}/Include CACHE STRING "")
set(FREERTOS_HEAP ${FW3_COMPONENTS_DIR}/Source/no_heap.c CACHE STRING "")
```

StevenClinardSpanIO committed:
FW3: Comment out cmake path template definitions

CMSIS-FreeRTOS's CMakeLists.txt requires some variables to be defined by the user.

The file includes uncommented example commands such as:
  set(FREERTOS_PORT "" CACHE STRING "FreeRTOS port name")
with the expectation that the user will set these variables either in this file
or by a parent.

FW3 will define these by the parent, so these statements need to be
removed so they will not replace the desired values.

StevenClinardSpanIO committed:
FW3: cmake set FreeRTOS symbols CACHE STRING ""

Setting FreeRTOS cmake symbols to cache string "" is required under Linux.
